README.md
## looker3 <a href="https://travis-ci.org/avantcredit/looker3"><img src="https://img.shields.io/travis/avantcredit/looker3.svg"></a> <a href="https://codecov.io/github/avantcredit/looker3"><img src="https://img.shields.io/codecov/c/github/avantcredit/looker3.svg"></a> <a href="https://github.com/avantcredit/looker3/tags"><img src="https://img.shields.io/github/tag/avantcredit/looker3.svg"></a>

Pull data from Looker with the 3.0 API.

## Installation

```R
if (!require("devtools")) { install.packages("devtools") }
devtools::install_github("avantcredit/looker3")
```

## How it works

Before pulling data from Looker, you need to set up environment variables LOOKER_URL, LOOKER_ID, and LOOKER_SECRET with the url to your Looker instance, your client id, and your client secret respectively.

Once those are set up, you can access data using the `looker3` function:
```R
library("looker3")
df <- pull_data(model = "thelook",
                view = "orders",
                fields = c("orders.count", "orders.created_month")
                filters = list("orders.created_month" = "90 days", "orders.status" = "complete")
)
```

# Specifying filters

There are two ways to specify filters.
* As a list, as illustrated above
* As a vector, using a colon as a separator, e.g.

```
filters = c("orders.created_month: 90 days", "orders.status: complete")
```


NEWS.md
# Version 0.2.1
- Updated unit tests (no changes in functionality)

# Version 0.2.0
- Renamed `looker3::looker3` to `pull_data`
- Added `get_sql` function, to request underlying sql query instead of data

# Version 0.1.15
- Error out if `looker3::looker3` is called with filters that duplicate a dimension name,
  since these will not be processed correctly (JSON objects cannot have duplicate names).

# Version 0.1.14
- Fixed bug where `validate_response` would crash if the response object was not a list.

# Version 0.1.13
- Added a warning if the number of columns returned is different from the number of fields provided.

# Version 0.1.12
- Added message from the http response from Looker to error messages coming from `validate_response`.

# Version 0.1.11
- Fixed bug where `looker3` was not passing `silent_read_csv` to `run_inline_query`.

# Version 0.1.10
- Added a param `silent_csv_read`, to switch off/on warnings while reading the stream into a data frame.

# Version 0.1.9
- Added an extra validation in `extract_query_response` to catch silent errors in the body of httr response objects.

# Version 0.1.8
- Also allow `filters` to be vector.

# Version 0.1.7
- `extract_query_result` now uses `readr::read_csv` instead of `utils::read.csv` under the hood.

# Version 0.1.6
- removed `logout_api_call`, using `cacher` to cache tokens instead.

# Version 0.1.5
- `validate_response` now only warns on unsuccessful logout validations.

# Version 0.1.4
- Error messages on `validate_response` now include step name and status code.

# Version 0.1.3
- Added compatibility for inputs in the format of Looker API 2.0.

# Version 0.1.0
- Added support for `run_inline_query` using the `csv` API endpoint(which supports streaming).

# Version 0.0.1
- Package created.


tests/testthat/test-looker3.R
context("pull_data")

fake_env_vars <- list(
  url    = "fake.looker.com:111/",
  id     = "fake_client",
  secret = "fake_secret"
)

withr::with_envvar(c(
  "LOOKER_URL"    = fake_env_vars$url,
  "LOOKER_ID"     = fake_env_vars$id,
  "LOOKER_SECRET" = fake_env_vars$secret
), {

  describe("handling missing env vars", {

    test_that("it stops if LOOKER_URL is missing", {
      withr::with_envvar(c("LOOKER_URL" = ""),
        expect_error(pull_data(model = "look", view = "items", fields = "name"),
          "place your Looker 3.0 API url in the environment")
      )
    })

    test_that("it stops if LOOKER_ID is missing", {
      withr::with_envvar(c("LOOKER_ID" = ""),
        expect_error(pull_data(model = "look", view = "items", fields = "name"),
          "place your Looker 3.0 client id in the environment")
      )
    })

    test_that("it stops if LOOKER_SECRET is missing", {
      withr::with_envvar(c("LOOKER_SECRET" = ""),
        expect_error(pull_data(model = "look", view = "items", fields = "name"),
          "place your Looker 3.0 client secret in the environment")
      )
    })
  })

  describe("pull_data passes arguments to run_inline_query correctly", {
    with_mock(
      `looker3:::run_inline_query` = function(url, id, secret,
                                             model, view, fields, filters,
                                             limit, ...) {
        list(url = url, id = id, secret = secret, model = model,
             view = view, fields = fields, filters = filters,
             limit = limit)
      }, {
      args <- list(model = "look", view = "items",
        fields = c("category.name", "products.count"),
        filters = list("category.name" = "socks"))

      bw_compatible_args <- list(model = "look", view = "items",
        fields = c("category.name", "products.count"),
        filters = "category.name: socks")

      test_that("it passes args correctly, with default limit 1000",
        expect_equal(do.call(pull_data, args),
         c(list(url = fake_env_vars$url, id = fake_env_vars$id, secret = fake_env_vars$secret),
           args,
           list(limit = 1000)))
      )
      test_that("it passes args correctly, with limit passed as an argument",
        expect_equal(do.call(pull_data, c(args, list(limit = 20))),
         c(list(url = fake_env_vars$url, id = fake_env_vars$id, secret = fake_env_vars$secret),
           args,
           list(limit = 20)))
      )
      test_that("it parses lookR-formatted filters correctly",
        expect_equal(do.call(pull_data, bw_compatible_args),
         c(list(url = fake_env_vars$url, id = fake_env_vars$id, secret = fake_env_vars$secret),
           args,
           list(limit = 1000)))
      )
      })
  })
})


tests/testthat/test-api_calls.R
context("api calling helper functions")

with_mock(
  `httr::POST` = function(...) NULL,
  `httr::DELETE` = function(...) NULL,
  `httr::add_headers` = function(Authorization) paste0("Authorization is ", Authorization), {

    test_that("login_api_call passes url to httr::POST", {
      with_mock(`httr::POST` = function(url) {
          identical(url,"https://fake.looker.com:111/api/3.0/login?client_id=FAKE_ID&client_secret=FAKE_SECRET")
        }, {
          expect_true(login_api_call("https://fake.looker.com:111/",
                        "FAKE_ID", "FAKE_SECRET"))
      })
    })

    with_mock(
    `looker3:::cached_token_is_invalid` = function(...) { FALSE },
    `httr::POST` = function(url, header, body, encode) {
      list(url = url, header = header, body = body)
    }, {
      args <- list(base_url = "https://fake.looker.com:111/",
        model = "look", view = "items",
        fields = c("category.name", "products.count"),
        filters = list(c("category.name", "socks")),
        return_format = "csv")

      token_cache$set("token", list(token = "FAKE_TOKEN"))

      test_that("query_api_call passes url to httr::POST", {
      expect_identical(
        do.call(query_api_call, args)$url,
        "https://fake.looker.com:111/api/3.0/queries/run/csv")
      })

      test_that("query_api_call passes url to httr::POST", {
      sql_args <- within(args, {return_format <- "sql"})
      expect_identical(
        do.call(query_api_call, sql_args)$url,
        "https://fake.looker.com:111/api/3.0/queries/run/sql")
      })

      test_that("query_api_call passes token to httr::POST", {
      expect_identical(
        do.call(query_api_call, args)$header,
          "Authorization is token FAKE_TOKEN")
      })

      test_that("query_api_call passes body to httr::POST", {
      expect_equal(
        do.call(query_api_call, args)$body,
        list(model = args$model, view = args$view, fields = I(args$fields),
             filters = args$filters, limit = 1000))
      })

      test_that("query_api_call passes user-specified limit to httr::POST", {
      expect_equal(
        do.call(query_api_call, c(args, list(limit = 20)))$body,
        list(model = args$model, view = args$view, fields = I(args$fields),
             filters = args$filters, limit = 20))
      })
    })
})


tests/testthat/test-run_inline_query.R
context("run_inline_query")

args <- list(
  base_url = "fake.looker.com/",
  client_id = "fake_client",
  client_secret = "fake_secret",
  model = "fake_model",
  view = "fake_view",
  fields = c("fake", "field"),
  filters = list(c("fake.name", "filtered"))
)

describe("run_inline_query helpers called with the corresponding inputs", {
  with_mock(
    `looker3:::login_api_call` = function(...) NULL,
    `looker3:::logout_api_call` = function(...) NULL,
    `looker3:::cached_token_is_invalid` = function(...) FALSE,
    `looker3:::query_api_call` = function(...) NULL,
    `looker3:::extract_data_from_response` = function(...) NULL,
    `looker3:::extract_sql_from_response` = function(...) NULL, {

      test_that("login_api_call called if cached token is invalid", {
        with_mock(
          `looker3:::login_api_call` = function(base_url, client_id, client_secret) {
            stop(paste("login_api_call:", base_url, client_id, client_secret))
          },
          `looker3:::cached_token_is_invalid` = function(...) { TRUE }, {
            expect_error(do.call(run_inline_query, args),
              "login_api_call: fake.looker.com/ fake_client fake_secret")
        })
      })

      test_that("login procedure bypassed if cached token is valid", {
        with_mock(
          `looker3:::login_api_call` = function(...) {
            stop("stopped by login_api_call")
          },
          `looker3:::put_new_token_in_cache` = function(...) {
            stop("stopped by put_new_token_in_cache")
          },
          `looker3:::query_api_call` = function(...) {
            stop("bypassed token caching")
          }, {
            expect_error(do.call(run_inline_query, args),
              "bypassed token caching")
        })
      })

      test_that("query_api_call receives its args, including token", {
        with_mock(
          `looker3:::query_api_call` = function(base_url, model, view, fields, filters, limit, ...) {
            actual_inputs <- list(base_url = base_url, model = model, view = view, fields = fields,
                           filters = filters, limit = limit)
            expected_inputs <- list(base_url = args$base_url, model = args$model, view = args$view,
                            fields = args$fields, filters = args$filters, limit = 1000)
            if (identical(actual_inputs, expected_inputs)) {
              stop("query_api_call called correctly")
            }
           stop("query_api_call with incorrect inputs")
          }, {
            expect_error(do.call(run_inline_query, args), "query_api_call called correctly")
          })

        with_mock(
          `looker3:::query_api_call` = function(base_url, model, view, fields, filters, limit, ...) {
            actual_inputs <- list(base_url = base_url, model = model, view = view, fields = fields,
                           filters = filters, limit = limit)
            expected_inputs <- list(base_url = args$base_url, model = args$model, view = args$view,
                            fields = args$fields, filters = args$filters, limit = 20)
            if (identical(actual_inputs, expected_inputs)) {
              stop("query_api_call called correctly")
            }
            stop("failure")
          }, {
            expect_error(do.call(run_inline_query, c(args, list(limit = 20)), "query_api_call called correctly"))
        })
      })

      test_that("return_format determines the correct extractor to receives the output of query_api_call", {
        with_mock(
          `looker3:::query_api_call` = function(...) "response received",
          `looker3:::extract_data_from_response` = function(inline_query_response, ...) {
            stop(paste0("extract_data_from_response called: "), inline_query_response)
          },
          `looker3:::extract_sql_from_response` = function(inline_query_response, ...) {
            stop(paste0("extract_sql_from_response called: "), inline_query_response)
          }, {
            csv_args <- within(args, {return_format = "csv"})
            sql_args <- within(args, {return_format = "sql"})
            expect_error(do.call(run_inline_query, csv_args),
              "extract_data_from_response called: response received")
            expect_error(do.call(run_inline_query, sql_args),
              "extract_sql_from_response called: response received")
        })
      })
  })
})


R/utils.R
`%||%` <- function(x, y) if (is.null(x)) y else x

colon_split_to_list <- function(string) {
  colon_split <- strsplit(string, ": ")
  field_names <- lapply(colon_split, `[[`, 1)
  values <- lapply(colon_split, `[[`, 2)
  names(values) <- field_names
  values
}

list_to_colon_split <- function(list) {
  paste(names(list), list, sep = ": ") %>% unlist
}

setup_using_env_vars <- function() {
  env_var_descriptions <- list(
    LOOKER_URL    = "API url",
    LOOKER_ID     = "client id",
    LOOKER_SECRET = "client secret"
  )

  looker_setup <- lapply(names(env_var_descriptions), function(name) {
    env_var <- Sys.getenv(name)
    if (env_var == "") {
      stop(paste0("Your environment variables are not set correctly. ",
        "please place your Looker 3.0 ", env_var_descriptions[[name]],
        " in the environment variable ", name, "."
      ))
    }
    env_var
  })

  names(looker_setup) <- names(env_var_descriptions)
  looker_setup
}


R/get_sql.R
#' Get underlying sql for a Looker query.
#'
#' This function assumes that you have set the environment
#' variables \code{LOOKER_URL}, \code{LOOKER_ID}, and
#' \code{LOOKER_SECRET}.
#'
#' The function will not allow the query to proceed unless you specify
#' a model, a view, and at least one field.
#'
#' @param model character. The \code{model} parameter of the query.
#' @param view character. The \code{view} parameter of the query.
#' @param fields character. The \code{fields} parameter of the query.
#' @param filters list or character.  Either a named list, or a character
#' vector using colons as separators
#' each vector describing one of the \code{filters} of the query.
#' @param limit numeric. The \code{limit} parameter of the query.
#'
#' @return a simple string containing the SQL query.
#'
#' @examples \dontrun{
#'   query <- get_sql(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = list("orders.created_month" = "90 days", "orders.status" = "complete")
#'   )
#'
#'   query <- get_sql(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = c("orders.created_month: 90 days", "orders.status: complete")
#'   )
#' }
#' @export
get_sql <- checkr::ensure(pre = list(   # model, view, and fields are
             model %is% simple_string,  # required to form a query.
             view %is% simple_string,
             fields %is% character,
             filters %is% simple_string ||
               ((filters %is% list || filters %is% vector) &&
                (filters %contains_only% simple_string || filters %is% empty)),
             limit %is% numeric && limit > 0 && limit %% 1 == 0
           ),

  function(model, view, fields,
            filters = list(), limit = 1000) {

    looker_setup <- setup_using_env_vars()

    # if user-specified filters as a character vector, reformat to a list
    if (!missing(filters) && is.character(filters)) {
      filters <- colon_split_to_list(filters)
    }

    run_inline_query(looker_setup$LOOKER_URL, looker_setup$LOOKER_ID, looker_setup$LOOKER_SECRET,
      model, view, fields, filters, limit, return_format = "sql", silent_read_csv)
  }
)


R/api_calls.R

login_api_call <- function(base_url, client_id, client_secret) {
  login_url <- paste0(base_url, "api/3.0/login?client_id=", client_id,
                "&client_secret=", client_secret)
  httr::POST(login_url)
}

logout_api_call <- function(base_url, token) {
  logout_url <- paste0(base_url, "api/3.0/logout")
  httr::DELETE(logout_url,
    httr::add_headers(Authorization = paste0("token ", token)))
}

query_api_call <- function(base_url, model, view, fields, filters,
                    limit = 1000, return_format) {
  token      <- token_cache$get("token")$token
  query_url  <- paste0(base_url, "api/3.0/queries/run/", return_format)
  query_body <- list(model = model, view = view, fields = I(fields),
                  filters = filters, limit = limit)
  if (length(filters) == 0) { query_body$filters <- NULL }

  httr::with_config(httr::timeout(getOption("looker3.timeout", 7200)),
    httr::POST(url = query_url,
      httr::add_headers(Authorization = paste0("token ", token)),
      body = query_body, encode = "json")
  )
}


R/pull_data.R
#' Pull data from Looker.
#'
#' This function assumes that you have set the environment
#' variables \code{LOOKER_URL}, \code{LOOKER_ID}, and
#' \code{LOOKER_SECRET}.
#'
#' The function will not allow the query to proceed unless you specify
#' a model, a view, and at least one field.
#'
#' @param model character. The \code{model} parameter of the query.
#' @param view character. The \code{view} parameter of the query.
#' @param fields character. The \code{fields} parameter of the query.
#' @param filters list or character.  Either a named list, or a character
#' vector using colons as separators
#' each vector describing one of the \code{filters} of the query.
#' @param limit numeric. The \code{limit} parameter of the query.
#' @param silent_read_csv logical. Whether or not to suppress warnings
#' when reading the streamed data into a data frame.
#'
#' @return a data.frame containing the data returned by the query
#'
#' @examples \dontrun{
#'   df <- pull_data(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = list("orders.created_month" = "90 days", "orders.status" = "complete")
#'   )
#'
#'   df <- pull_data(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = c("orders.created_month: 90 days", "orders.status: complete")
#'   )
#' }
#' @export
pull_data <- checkr::ensure(pre = list(   # model, view, and fields are
             model %is% simple_string,  # required to form a query.
             view %is% simple_string,
             fields %is% character,
             filters %is% simple_string ||
               ((filters %is% list || filters %is% vector) &&
                (filters %contains_only% simple_string || filters %is% empty)),
             limit %is% numeric && limit > 0 && limit %% 1 == 0,
             silent_read_csv %is% logical
           ),

  function(model, view, fields,
            filters = list(), limit = 1000,
            silent_read_csv = TRUE) {

    looker_setup <- setup_using_env_vars()

    # if user-specified filters as a character vector, reformat to a list
    if (!missing(filters) && is.character(filters)) {
      filters <- colon_split_to_list(filters)
      if (anyDuplicated(names(filters))) {
        stop("Looker query contains duplicate filters and only the first one ",
             "will be executed. Please drop or coalesce these: ",
             paste(unique(names(filters)[duplicated(names(filters))]), collapse = ", "),
             call. = FALSE)
      }
    }


    run_inline_query(looker_setup$LOOKER_URL, looker_setup$LOOKER_ID, looker_setup$LOOKER_SECRET,
      model, view, fields, filters, limit, return_format = "csv", silent_read_csv)
  }
)



R/pull_query.R
#' Get underlying sql for a Looker query.
#'
#' This function assumes that you have set the environment
#' variables \code{LOOKER_URL}, \code{LOOKER_ID}, and
#' \code{LOOKER_SECRET}.
#'
#' The function will not allow the query to proceed unless you specify
#' a model, a view, and at least one field.
#'
#' @param model character. The \code{model} parameter of the query.
#' @param view character. The \code{view} parameter of the query.
#' @param fields character. The \code{fields} parameter of the query.
#' @param filters list or character.  Either a named list, or a character
#' vector using colons as separators
#' each vector describing one of the \code{filters} of the query.
#' @param limit numeric. The \code{limit} parameter of the query.
#'
#' @return a simple string containing the SQL query.
#'
#' @examples \dontrun{
#'   df <- get_sql(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = list("orders.created_month" = "90 days", "orders.status" = "complete")
#'   )
#'
#'   df <- get_sql(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = c("orders.created_month: 90 days", "orders.status: complete")
#'   )
#' }
#' @export

get_sql <- checkr::ensure(pre = list(   # model, view, and fields are
             model %is% simple_string,  # required to form a query.
             view %is% simple_string,
             fields %is% character,
             filters %is% simple_string ||
               ((filters %is% list || filters %is% vector) &&
                (filters %contains_only% simple_string || filters %is% empty)),
             limit %is% numeric && limit > 0 && limit %% 1 == 0
           ),

  function(model, view, fields,
            filters = list(), limit = 1000) {

    looker_setup <- setup_using_env_vars()

    # if user-specified filters as a character vector, reformat to a list
    if (!missing(filters) && is.character(filters)) {
      filters <- colon_split_to_list(filters)
    }

    run_inline_query(looker_setup$LOOKER_URL, looker_setup$LOOKER_ID, looker_setup$LOOKER_SECRET,
      model, view, fields, filters, limit, return_format = "sql")
  }
)


R/run_inline_query.R
#' Run an inline query using the Looker 3.0 API.
#'
#' @param base_url character. The base url of the Looker server.
#' @param client_id character. Your client id.
#' @param client_secret character. Your client secret.
#' @param model character. The \code{model} parameter of the query.
#' @param view character. The \code{view} parameter of the query.
#' @param fields character. The \code{fields} parameter of the query.
#' @param filters named list. The names and entries of the list will
#' be converted into a hash table and passed as the \code{filters}
#' of the query.
#' @param limit numeric. The \code{limit} parameter of the query.
#'
#' @return a data.frame containing the data returned by the query.
#'
run_inline_query <- function(base_url, client_id, client_secret,
                      model, view, fields, filters,
                      limit = 1000, return_format = c("csv", "sql"), silent_read_csv) {

  return_format <- match.arg(return_format)

  # The API requires you to have an access token obtained by "logging in".
  if (cached_token_is_invalid()) {
    login_response <- login_api_call(base_url, client_id, client_secret)
    put_new_token_in_cache(login_response)
  }

  inline_query_response <- query_api_call(base_url,
    model, view, fields, filters, limit, return_format)



  if (identical(return_format, "csv")) {
    output   <- extract_data_from_response(inline_query_response, silent_read_csv)
    colnames <- names(output)

    if (length(colnames) != length(fields)) {
      warning(paste("Number of columns returned doesn't match number of fields provided.",
        "\n Looker data column names:", paste0(colnames, collapse = ", "),
        "\n fields provided:", paste0(list_to_colon_split(fields), collapse = ", "))
      )
    }
  }


  if (identical(return_format, "sql")) {
    output <- extract_sql_from_response(inline_query_response)
  }

  output
}


R/response_handlers.R
cached_token_is_invalid <- function() {
  if (!token_cache$exists("token")) { return(TRUE) }
  expiration_date <- token_cache$get("token")$expires_in
  is.null(expiration_date) || methods::is(expiration_date, "POSIXt") || i
    (expiration_date < Sys.time())
}

is.successful_response <- function(response) {
  httr::status_code(response) %in% c("200", "201", "202", "204")
}

extract_error_message <- function(response) {
  response_body <- httr::content(response)
  if (is.list(response_body)) {
    response_body$message
  } else {
    response_body
  }
}

validate_response <- function(response) {
  if (is.successful_response(response)) { return(TRUE) }

  stop(paste("The",
    gsub("_", " ", deparse(substitute(response))),
    "step in looker failed. The status code was",
    httr::status_code(response),
    "and the error message was",
    extract_error_message(response) %||% "not provided."
    )
  )
}

put_new_token_in_cache <- function(login_response) {
  validate_response(login_response)
  token_cache$set("token", list(
    token      = httr::content(login_response)$access_token,
    # avoid token expiration during code execution
    expires_in = Sys.time() + httr::content(login_response)$expires_in - 1 
  ))
}

extract_data_from_response <- function(query_response, silent_read_csv = TRUE) {
  validate_response(query_response)
  data_from_query <- httr::content(query_response)
  if (grepl("^Error:", data_from_query)) {
    # assume that the query errored quietly and that data_from_query is an error message.
    stop("Looker returned no data and the following error message:\n", as.character(data_from_query))
  }
  if (silent_read_csv) {
    suppressWarnings(readr::read_csv(data_from_query))
  } else {
    readr::read_csv(data_from_query)
  }
}

extract_sql_from_response <- function(query_response) {
  validate_response(query_response)
  httr::content(query_response)
}



R/looker3.R
#' Pull data from Looker.
#'
#' This function assumes that you have set the environment
#' variables \code{LOOKER_URL}, \code{LOOKER_ID}, and 
#' \code{LOOKER_SECRET}.
#'
#' The function will not allow the query to proceed unless you specify
#' a model, a view, and at least one field.
#'
#' @param model character. The \code{model} parameter of the query.
#' @param view character. The \code{view} parameter of the query.
#' @param fields character. The \code{fields} parameter of the query.
#' @param filters list or character.  Either a named list, or a character
#' vector using colons as separators
#' each vector describing one of the \code{filters} of the query.
#' @param limit numeric. The \code{limit} parameter of the query.
#' @param silent_read_csv logical. Whether or not to suppress warnings
#' when reading the streamed data into a data frame.
#'
#' @return a data.frame containing the data returned by the query
#'
#' @examples \dontrun{
#'   df <- looker3(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = list("orders.created_month" = "90 days", "orders.status" = "complete")
#'   )
#' 
#'   df <- looker3(model = "thelook",
#'                 view = "orders",
#'                 fields = c("orders.count", "orders.created_month")
#'                 filters = c("orders.created_month: 90 days", "orders.status: complete")
#'   )  
#' }
#' @export
looker3 <- checkr::ensure(pre = list(   # model, view, and fields are
             model %is% simple_string,  # required to form a query.
             view %is% simple_string,
             fields %is% character,
             filters %is% simple_string ||
               ((filters %is% list || filters %is% vector) &&
                (filters %contains_only% simple_string || filters %is% empty)),
             limit %is% numeric && limit > 0 && limit %% 1 == 0,
             silent_read_csv %is% logical
           ), 

  function(model, view, fields,
            filters = list(), limit = 1000,
            silent_read_csv = TRUE) {

    env_var_descriptions <- list(
      LOOKER_URL    = "API url",
      LOOKER_ID     = "client id",
      LOOKER_SECRET = "client secret"
    )

    looker_setup <- lapply(names(env_var_descriptions), function(name) {
      env_var <- Sys.getenv(name)
      if (env_var == "") {
        stop(paste0("Your environment variables are not set correctly. ",
          "please place your Looker 3.0 ", env_var_descriptions[[name]],
          " in the environment variable ", name, "."
        ))
      }
      env_var
    })

    names(looker_setup) <- names(env_var_descriptions)

    # if user-specified filters as a character vector, reformat to a list
    if (!missing(filters) && is.character(filters)) {
      filters <- colon_split_to_list(filters) 
      # Take duplicate filters which will have multiple occurences of the same key
      # name in the filters list and flatten them into character vectors so each
      # key is unique.
      filters <- tapply(filters, names(filters), FUN = function(queries) { unlist(unname(queries)) })
    }


    run_inline_query(looker_setup$LOOKER_URL, looker_setup$LOOKER_ID, looker_setup$LOOKER_SECRET,
      model, view, fields, filters, limit, silent_read_csv)
  }
)

